### ISSUE TITLE
Add front-end stack

### ISSUE THREAD

**Issue by jmaupetit** [Candihub/pixel#10]
## Purpose

We need to setup a front-end stack to ease app styling and rich UI development.

## Proposal

* Set up a base node8/yarn environment
* Use Sass with the Foundation framework
* Use webpack to built assets

**Comment by pierrepo** [comment]
The Bootstrap framework could not do the job?
What are node8/yarn, Sass and webpack?

**Comment by jmaupetit** [comment]
Sorry for the front-end jargon, here is a more detailed list of tools we plan to use:

* Node 8: latest release of  [NodeJS](https://nodejs.org/en/)
* [`yarn`](https://yarnpkg.com/lang/en/): a package manager for node (it's the new `npm`)
* [Sass](http://sass-lang.com): Syntactically awesome style sheets — a css pre-processor to ease CSS development
* [webpack](https://webpack.js.org/): a module bundler for JavaScript applications

For now [Twitter Bootstrap](http://getbootstrap.com/) framework is still less extensible compared to the [Foundation](https://foundation.zurb.com/sites/docs/) framework; mostly because each Foundation component is accessible via a [Sass Mixin](http://sass-lang.com/guide#topic-6), offering the opportunity to develop a semantic DOM. Twitter bootstrap almost forces you to use its CSS class names or [`@extend`](http://sass-lang.com/guide#topic-7) them.

**Comment by pierrepo** [comment]
OK. Thanks. But why not getting directly the [default Foundation CSS file](https://foundation.zurb.com/sites/docs/installation.html#css-download)? Afterwards, we might customize a bit the CSS if needed (yes, by hand).
I still do not get the advantage of getting NodeJS+yarn+Sass+webpack...

**Comment by jmaupetit** [comment]
Okay, I will try to clarify this with an example.

Let's say that you have the following class in Foundation's default CSS file:

```css
/* foundation.css */

.button {
  margin: 0.2rem;
  padding: 0.7rem 0.3rem;
  background-color: red;
  border-radius: 5px;
  color: white;
  font-size: 1.2rem;
}
```

And you want to use this class for a link:

```html
<a href="#" class="button">super cool link</a>
```

But the background color does not suite your design, so you'll want to override default color with a custom class written by-hand:

```css
/* my-styles.css */

.action-button {
  background-color: blue;
}
```

A CSS class that you will need to add in the DOM:

```html
<a href="#" class="button action-button">super cool link</a>
```

And your browser will compute this link style as:

```css
.button.action-button {
  margin: 0.2rem;
  padding: 0.7rem 0.3rem;
  /* the web browser overrides this css property as the second CSS class modifies it */
  /* background-color: red; */ 
  background-color: blue;
  border-radius: 5px;
  color: white;
  font-size: 1.2rem;
}
```

In this case, you have a non-semantic DOM (`a.button.action-button` instead of `a.action-button`), and overriding CSS properties by stacking classes increases page load time (because your web browser has more things to compute).

Using Sass improves the situation: let's say we want to use foundation's `button` mixin defined as follow:

```scss
/* foundation/_button.scss */

@mixin button($bg: red) {
  margin: 0.2rem;
  padding: 0.7rem 0.3rem;
  background-color: $bg;
  border-radius: 5px;
  color: white;
  font-size: 1.2rem;
}
```

We have to load foundation's mixins and customize a button style that suit our needs as follow:

```scss
/* project.scss */
@import 'foundation/button';

.action-button {
  @include button($bg: blue);
}
```

This example above will compile into this CSS:

```css
.action-button {
  margin: 0.2rem;
  padding: 0.7rem 0.3rem;
  background-color: blue;
  border-radius: 5px;
  color: white;
  font-size: 1.2rem; 
}
```

Our DOM is cleaner, our styles are smaller. Smart isn't it? 😎 

This is a small example, but Sass offers way more features to programmatically handle your styles.

Moreover, consider this huge blob of CSS classes that will be loaded and never used in the default Foundation's styles!

**Comment by pierrepo** [comment]
OK, but I thought the purpose of CSS was *Cascading*?

> Moreover, consider this huge blob of CSS classes that will be loaded and never used in the default Foundation's styles!

Do you mean that with Sass we will be able to generate a css with only the properties/features of Foundation we want to use? Sounds interesting indeed.

**Comment by jmaupetit** [comment]
> OK, but I thought the purpose of CSS was Cascading?

Good point 😉 but it does not mean this will be optimal for page loading time, etc.

> Do you mean that with Sass we will be able to generate a css with only the properties/features of Foundation we want to use? Sounds interesting indeed.

Yes! You are free to only import components you will use via the `@import` keyword, _e.g._
 `@import 'foundation/button';`

**Comment by pierrepo** [comment]
OK. Make me dream! :sunglasses: 
But please document as much as you can what you are doing (and why and how) :pencil: :pencil: :pencil: :pencil: 
I do really fear the curse of dependencies :scream:

**Comment by jmaupetit** [comment]
I understand your concern. We will try to ease as much as possible project bootstrapping and make it rock-solid & reproducible 💪

**Comment by jmaupetit** [comment]
Fixed in #36
